### PR TITLE
Make cache dir configurable

### PR DESCRIPTION
**Pull request by Bond-009**

### CODE REVIEW

File: Jellyfin.Server/Program.cs [hunk 1]
```diff
@@ -187,6 +187,30 @@ private static ServerApplicationPaths CreateApplicationPaths(StartupOptions opti
...
+            if (string.IsNullOrEmpty(cacheDir))
+            {
+                if (options.ContainsOption("-cachedir"))
+                {
+                    cacheDir = options.GetOption("-cachedir");
+                }
+                else if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                {
+                    // $XDG_CACHE_HOME defines the base directory relative to which user specific non-essential data files should be stored.
+                    cacheDir = Environment.GetEnvironmentVariable("XDG_CACHE_HOME");
```
Comment (ploughpuff) [inline Jellyfin.Server/Program.cs]: Why no 'jellyfin' folder under XDG_CACHE_HOME?

File: Jellyfin.Server/Program.cs [hunk 2]
```diff
@@ -187,6 +187,30 @@ private static ServerApplicationPaths CreateApplicationPaths(StartupOptions opti
...
+                    // If $XDG_CACHE_HOME is either not set or empty, $HOME/.cache should be used.
+                    if (string.IsNullOrEmpty(cacheDir))
+                    {
+                        cacheDir = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".cache", "jellyfin");
+                    }
+                }
```
Comment (ploughpuff) [inline Jellyfin.Server/Program.cs]: No else{} for when running on windows.  I guess in that case we use the code in BaseApplicationPath CachePath 'get' to create the path to <datadir>/cache (yuck for me).  Might it be clear if that logic was moved here now, then getter becomes cleaner?

**Comment by ploughpuff** [comment]
In addition, the new cacheDir variable is not being used in the call to ServerApplicationPaths():
`return new ServerApplicationPaths(programDataPath, appPath, appPath, logDir, configDir);`

This should probably be:
`return new ServerApplicationPaths(programDataPath, appPath, appPath, logDir, configDir, cacheDir);`

File: Jellyfin.Server/Program.cs (hunk 1 above)
Comment (Bond-009) [reply Jellyfin.Server/Program.cs]: There is.

Comment (ploughpuff) [reply Jellyfin.Server/Program.cs]: Only if XDG_CACHE_HOME is null do we drop to using HOME/.cache/ and append the 'jellyfin'.  If XDG_CACHE_HOME is valid, then no 'jellyfin' is not appended.

File: Jellyfin.Server/Program.cs (hunk 2 above)
Comment (Bond-009) [reply Jellyfin.Server/Program.cs]: Not a good idea, `<datadir>/cache` is the default for all implementations now, moving it out will force each separate implementation (ex: Mac app) to properly set up all paths.

**Comment by ploughpuff** [comment]
Slightly related, can you take a look at UpdateCachePath().  It's called during init (new CoreAppHost) and seems to want to set the cache path to null.  I think this fails (to set null) because the storage is private and app continues to load with correct set path:
```
private string _cachePath;
set => _cachePath = value;
```

**Comment by joshuaboniface** [comment]
Looks good to me - going to modify the various build facilities to make use of this, then merge it.

**Comment by joshuaboniface** [comment]
@Bond-009 Can you update this to match the results of #749, including help text.

**Comment by Bond-009** [comment]
@joshuaboniface Done, don't know if the short arg is good tho.

File: Jellyfin.Server/StartupOptions.cs [hunk 3]
```diff
@@ -11,6 +11,9 @@ public class StartupOptions : IStartupOptions
         [Option('d', "datadir", Required = false, HelpText = "Path to use for the data folder (database files, etc.).")]
         public string DataDir { get; set; }
 
+        [Option('C', "configdir", Required = false, HelpText = "Path to use for caching.")]
```
Comment (ploughpuff) [inline Jellyfin.Server/StartupOptions.cs]: --configdir should be --cachedir?

Comment (ploughpuff) [reply Jellyfin.Server/StartupOptions.cs]: Not sure about the use of capital C.  Maybe not quick quick option for cache..

Comment (joshuaboniface) [reply Jellyfin.Server/StartupOptions.cs]: Capital C is a fine convention IMO.

**Pull request merged**